File: src/overlay.zig
// overlay: FAT32 permission-overlay parser.
//
// Pure, allocation-free, no externs. FAT32 has no native owner/mode
// concept, so /mnt files get their permission metadata from a root-level
// text file (/mnt/PERMS.TAB) instead of the 8d hard default.
// src/fat32_backend.zig reads that file once at mount time, parses it with
// these helpers into a fixed table, and open() consults the table;
// un-annotated paths keep the documented default (0666 root:root, except
// the shadow basename, which floors at 0600 — see fat32_backend.open).
// The overlay protects itself through its own entry (`PERMS.TAB 0600 0 0`).
//
// Line format: `NAME MODE UID GID`
//   * NAME — 8.3 basename as it appears in the FAT32 root; matched
//     case-insensitively (FAT32 names are caseless)
//   * MODE — octal permission word, low 9 bits only (no file-type bits;
//     the backend ORs the regular-file type back in)
//   * UID / GID — decimal
//   * `#` starts a comment; blank lines are skipped; CRLF tolerated
//     (cards get hand-edited on host machines)
//
// parse() rejects a malformed overlay WHOLESALE (returns null) instead of
// skipping bad lines: a half-applied policy is indistinguishable from a
// truncated or corrupted file, and the backend's corruption response (loud
// boot message + shadow floor) must fire for those, not silently shrink
// the table. The truth table below is the stage gate: no backend wiring
// ships until every row passes.

pub const MAX_ENTRIES: usize = 16;
// 8.3 basename: 8 name chars + '.' + 3 extension chars.
pub const MAX_NAME: usize = 12;

pub const Entry = struct {
    name_buf: [MAX_NAME]u8,
    name_len: u8,
    mode: u32,
    uid: u32,
    gid: u32,

    pub fn name(self: *const Entry) []const u8 {
        return self.name_buf[0..self.name_len];
    }
};

// Parse the overlay text into `out`. Returns the entry count (0 for an
// empty or comment-only file — valid), or null on the first malformed
// line: missing field, 5th field, non-octal mode, mode above 0o777,
// non-decimal uid/gid, empty or over-long name, or more entries than
// `out` holds.
pub fn parse(content: []const u8, out: []Entry) ?usize {
    var count: usize = 0;
    var line_start: usize = 0;
    var i: usize = 0;
    while (i <= content.len) : (i += 1) {
        if (i != content.len and content[i] != '\n') continue;
        var line = content[line_start..i];
        line_start = i + 1;
        // CRLF tolerance: strip one trailing carriage return.
        if (line.len > 0 and line[line.len - 1] == '\r') line = line[0 .. line.len - 1];
        const trimmed = trim(line);
        if (trimmed.len == 0) continue;
        if (trimmed[0] == '#') continue;

        // Split into exactly 4 whitespace-separated fields.
        var fields: [4][]const u8 = undefined;
        var nf: usize = 0;
        var j: usize = 0;
        while (j < trimmed.len) {
            while (j < trimmed.len and isSpace(trimmed[j])) j += 1;
            if (j >= trimmed.len) break;
            const fstart = j;
            while (j < trimmed.len and !isSpace(trimmed[j])) j += 1;
            if (nf == 4) return null; // a 5th field is malformed
            fields[nf] = trimmed[fstart..j];
            nf += 1;
        }
        if (nf != 4) return null;

        const fname = fields[0];
        if (fname.len == 0 or fname.len > MAX_NAME) return null;
        const mode = parseOctalU32(fields[1]) orelse return null;
        if (mode > 0o777) return null;
        const uid = parseDecimalU32(fields[2]) orelse return null;
        const gid = parseDecimalU32(fields[3]) orelse return null;

        if (count == out.len) return null; // over capacity
        out[count] = .{
            .name_buf = undefined,
            .name_len = @intCast(fname.len),
            .mode = mode,
            .uid = uid,
            .gid = gid,
        };
        for (fname, 0..) |c, k| out[count].name_buf[k] = c;
        count += 1;
    }
    return count;
}

// Case-insensitive lookup of `name_in` among `entries` (pass the parsed
// prefix, e.g. table[0..count]). First match wins.
pub fn lookup(entries: []const Entry, name_in: []const u8) ?Entry {
    for (entries) |e| {
        if (nameEql(e.name(), name_in)) return e;
    }
    return null;
}

// FAT32 name equality: case-insensitive byte comparison (8.3 names are
// caseless). Public so the backend can apply the same rule to names that
// are not in the table (the shadow floor check).
pub fn nameEql(a: []const u8, b: []const u8) bool {
    return eqlIgnoreCase(a, b);
}

fn isSpace(c: u8) bool {
    return c == ' ' or c == '\t';
}

fn trim(s: []const u8) []const u8 {
    var start: usize = 0;
    var end: usize = s.len;
    while (start < end and isSpace(s[start])) start += 1;
    while (end > start and isSpace(s[end - 1])) end -= 1;
    return s[start..end];
}

fn toLower(c: u8) u8 {
    return if (c >= 'A' and c <= 'Z') c + ('a' - 'A') else c;
}

fn eqlIgnoreCase(a: []const u8, b: []const u8) bool {
    if (a.len != b.len) return false;
    for (a, b) |x, y| {
        if (toLower(x) != toLower(y)) return false;
    }
    return true;
}

// Octal u32 parse, exact (digits 0-7 only, no 0o prefix, no sign).
fn parseOctalU32(s: []const u8) ?u32 {
    if (s.len == 0) return null;
    var v: u64 = 0;
    for (s) |c| {
        if (c < '0' or c > '7') return null;
        v = v * 8 + (c - '0');
        if (v > 0xFFFF_FFFF) return null;
    }
    return @intCast(v);
}

// Decimal u32 parse, exact (no sign, no whitespace).
fn parseDecimalU32(s: []const u8) ?u32 {
    if (s.len == 0) return null;
    var v: u64 = 0;
    for (s) |c| {
        if (c < '0' or c > '9') return null;
        v = v * 10 + (c - '0');
        if (v > 0xFFFF_FFFF) return null;
    }
    return @intCast(v);
}

// ---- Host tests ----
//
// The truth table below is the gate for the FAT32 overlay: the backend
// wiring (fat32_backend.applyOverlay / open lookup) does not ship until
// every row passes. Rows pin the format the seed file
// (user_space/etc/perms.tab) and the deploy/make_test_disk seeding use.
const std = @import("std");
const testing = std.testing;

test "parse: well-formed multi-line overlay" {
    var table: [MAX_ENTRIES]Entry = undefined;
    const content =
        "PERMS.TAB 0600 0 0\n" ++
        "SHADOW 0600 0 0\n" ++
        "ROUNDTR.DAT 0666 0 0\n";
    const n = parse(content, &table).?;
    try testing.expectEqual(@as(usize, 3), n);
    try testing.expectEqualStrings("PERMS.TAB", table[0].name());
    try testing.expectEqual(@as(u32, 0o600), table[0].mode);
    try testing.expectEqual(@as(u32, 0), table[0].uid);
    try testing.expectEqual(@as(u32, 0), table[0].gid);
    try testing.expectEqualStrings("SHADOW", table[1].name());
    try testing.expectEqual(@as(u32, 0o666), table[2].mode);
}

test "parse: comments, blank lines, and surrounding whitespace are skipped" {
    var table: [MAX_ENTRIES]Entry = undefined;
    const content =
        "# FlashOS FAT32 permission overlay\n" ++
        "\n" ++
        "   \n" ++
        "  SHADOW   0600  0   0  \n" ++
        "# trailing comment\n";
    const n = parse(content, &table).?;
    try testing.expectEqual(@as(usize, 1), n);
    try testing.expectEqualStrings("SHADOW", table[0].name());
}

test "parse: CRLF line endings are tolerated" {
    var table: [MAX_ENTRIES]Entry = undefined;
    const content = "SHADOW 0600 0 0\r\nPERMS.TAB 0600 0 0\r\n";
    const n = parse(content, &table).?;
    try testing.expectEqual(@as(usize, 2), n);
    try testing.expectEqualStrings("SHADOW", table[0].name());
    try testing.expectEqualStrings("PERMS.TAB", table[1].name());
}

test "parse: no trailing newline on the last line still parses" {
    var table: [MAX_ENTRIES]Entry = undefined;
    const n = parse("SHADOW 0600 0 0", &table).?;
    try testing.expectEqual(@as(usize, 1), n);
}

test "parse: empty and comment-only files are valid with zero entries" {
    var table: [MAX_ENTRIES]Entry = undefined;
    try testing.expectEqual(@as(usize, 0), parse("", &table).?);
    try testing.expectEqual(@as(usize, 0), parse("# nothing here\n", &table).?);
}

test "parse: missing field rejects the whole overlay" {
    var table: [MAX_ENTRIES]Entry = undefined;
    try testing.expectEqual(@as(?usize, null), parse("SHADOW 0600 0\n", &table));
    try testing.expectEqual(@as(?usize, null), parse("SHADOW 0600\n", &table));
    try testing.expectEqual(@as(?usize, null), parse("SHADOW\n", &table));
}

test "parse: a 5th field rejects the whole overlay" {
    var table: [MAX_ENTRIES]Entry = undefined;
    try testing.expectEqual(@as(?usize, null), parse("SHADOW 0600 0 0 extra\n", &table));
}

test "parse: one malformed line rejects the whole overlay (no partial table)" {
    var table: [MAX_ENTRIES]Entry = undefined;
    const content =
        "SHADOW 0600 0 0\n" ++
        "PERMS.TAB 9999 0 0\n"; // 9 is not an octal digit
    try testing.expectEqual(@as(?usize, null), parse(content, &table));
}

test "parse: non-octal mode and mode above 0777 reject" {
    var table: [MAX_ENTRIES]Entry = undefined;
    try testing.expectEqual(@as(?usize, null), parse("SHADOW 08 0 0\n", &table));
    try testing.expectEqual(@as(?usize, null), parse("SHADOW abc 0 0\n", &table));
    try testing.expectEqual(@as(?usize, null), parse("SHADOW 1777 0 0\n", &table));
}

test "parse: non-decimal uid / gid rejects" {
    var table: [MAX_ENTRIES]Entry = undefined;
    try testing.expectEqual(@as(?usize, null), parse("SHADOW 0600 root 0\n", &table));
    try testing.expectEqual(@as(?usize, null), parse("SHADOW 0600 0 0x0\n", &table));
}

test "parse: empty or over-long name rejects" {
    var table: [MAX_ENTRIES]Entry = undefined;
    // 13 chars: one past the 8.3 maximum.
    try testing.expectEqual(@as(?usize, null), parse("ABCDEFGHI.TXT 0600 0 0\n", &table));
}

test "parse: more entries than the table holds rejects" {
    var small: [2]Entry = undefined;
    const content =
        "A 0600 0 0\n" ++
        "B 0600 0 0\n" ++
        "C 0600 0 0\n";
    try testing.expectEqual(@as(?usize, null), parse(content, &small));
}

test "lookup: case-insensitive hit and miss" {
    var table: [MAX_ENTRIES]Entry = undefined;
    const n = parse("SHADOW 0600 0 0\nPERMS.TAB 0600 0 0\n", &table).?;
    // The backend looks paths up by their lowercase /mnt basename.
    const hit = lookup(table[0..n], "shadow").?;
    try testing.expectEqual(@as(u32, 0o600), hit.mode);
    const hit2 = lookup(table[0..n], "perms.tab").?;
    try testing.expectEqualStrings("PERMS.TAB", hit2.name());
    try testing.expectEqual(@as(?Entry, null), lookup(table[0..n], "roundtr.dat"));
}

test "lookup: empty table always misses" {
    const empty: [0]Entry = .{};
    try testing.expectEqual(@as(?Entry, null), lookup(&empty, "shadow"));
}

File: crates/kernel/src/fs/overlay.rs
//! FAT32 permission-overlay parser.
//!
//! FAT32 has no native owner or mode fields, so the root-level `PERMS.TAB`
//! supplies fixed metadata for selected 8.3 basenames. Parsing is allocation
//! free and rejects a malformed file wholesale: a partial policy is
//! indistinguishable from corruption, and the backend must fall back to its
//! protected defaults instead.

pub const MAX_ENTRIES: usize = 16;
pub const MAX_NAME: usize = 12;

/// One parsed `NAME MODE UID GID` row retained in the backend's fixed table.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[repr(C)]
pub struct Entry {
    pub name_buf: [u8; MAX_NAME],
    pub name_len: u8,
    pub mode: u32,
    pub uid: u32,
    pub gid: u32,
}

const _: () = assert!(core::mem::size_of::<Entry>() == 28);
const _: () = assert!(core::mem::align_of::<Entry>() == 4);
const _: () = assert!(core::mem::offset_of!(Entry, name_buf) == 0);
const _: () = assert!(core::mem::offset_of!(Entry, name_len) == 12);
const _: () = assert!(core::mem::offset_of!(Entry, mode) == 16);
const _: () = assert!(core::mem::offset_of!(Entry, uid) == 20);
const _: () = assert!(core::mem::offset_of!(Entry, gid) == 24);

impl Entry {
    pub fn name(&self) -> &[u8] {
        &self.name_buf[..usize::from(self.name_len)]
    }
}

/// Parse an overlay into `out`, returning the populated prefix length.
///
/// Empty and comment-only files are valid. Any malformed non-comment line or
/// capacity overflow rejects the complete input.
pub fn parse(content: &[u8], out: &mut [Entry]) -> Option<usize> {
    let mut count = 0usize;
    let mut line_start = 0usize;
    let mut i = 0usize;
    while i <= content.len() {
        if i == content.len() || content[i] == b'\n' {
            let mut line = &content[line_start..i];
            line_start = i.saturating_add(1);
            if line.last() == Some(&b'\r') {
                line = &line[..line.len() - 1];
            }
            let trimmed = trim(line);
            if !trimmed.is_empty() && trimmed[0] != b'#' {
                let mut fields = [&[][..]; 4];
                let mut field_count = 0usize;
                let mut cursor = 0usize;
                while cursor < trimmed.len() {
                    while cursor < trimmed.len() && is_space(trimmed[cursor]) {
                        cursor += 1;
                    }
                    if cursor >= trimmed.len() {
                        break;
                    }
                    let start = cursor;
                    while cursor < trimmed.len() && !is_space(trimmed[cursor]) {
                        cursor += 1;
                    }
                    if field_count == fields.len() {
                        return None;
                    }
                    fields[field_count] = &trimmed[start..cursor];
                    field_count += 1;
                }
                if field_count != fields.len() {
                    return None;
                }

                let name = fields[0];
                if name.is_empty() || name.len() > MAX_NAME {
                    return None;
                }
                let mode = parse_octal_u32(fields[1])?;
                if mode > 0o777 {
                    return None;
                }
                let uid = parse_decimal_u32(fields[2])?;
                let gid = parse_decimal_u32(fields[3])?;
                let slot = out.get_mut(count)?;
                let mut name_buf = [0; MAX_NAME];
                name_buf[..name.len()].copy_from_slice(name);
                *slot = Entry {
                    name_buf,
                    name_len: name.len() as u8,
                    mode,
                    uid,
                    gid,
                };
                count += 1;
            }
        }
        i += 1;
    }
    Some(count)
}

/// Return the first case-insensitive basename match.
pub fn lookup(entries: &[Entry], name: &[u8]) -> Option<Entry> {
    entries
        .iter()
        .copied()
        .find(|entry| name_eql(entry.name(), name))
}

/// FAT32 8.3 names compare case-insensitively over ASCII bytes.
pub fn name_eql(a: &[u8], b: &[u8]) -> bool {
    a.len() == b.len()
        && a.iter()
            .zip(b)
            .all(|(&left, &right)| lower(left) == lower(right))
}

fn is_space(byte: u8) -> bool {
    byte == b' ' || byte == b'\t'
}

fn trim(mut bytes: &[u8]) -> &[u8] {
    while bytes.first().is_some_and(|&byte| is_space(byte)) {
        bytes = &bytes[1..];
    }
    while bytes.last().is_some_and(|&byte| is_space(byte)) {
        bytes = &bytes[..bytes.len() - 1];
    }
    bytes
}

fn lower(byte: u8) -> u8 {
    if byte.is_ascii_uppercase() {
        byte + (b'a' - b'A')
    } else {
        byte
    }
}

fn parse_octal_u32(bytes: &[u8]) -> Option<u32> {
    parse_u32(bytes, 8, b'7')
}

fn parse_decimal_u32(bytes: &[u8]) -> Option<u32> {
    parse_u32(bytes, 10, b'9')
}

fn parse_u32(bytes: &[u8], radix: u64, max_digit: u8) -> Option<u32> {
    if bytes.is_empty() {
        return None;
    }
    let mut value = 0u64;
    for &byte in bytes {
        if byte < b'0' || byte > max_digit {
            return None;
        }
        value = value * radix + u64::from(byte - b'0');
        if value > u64::from(u32::MAX) {
            return None;
        }
    }
    Some(value as u32)
}

#[cfg(test)]
mod tests {
    use super::{lookup, parse, Entry, MAX_ENTRIES};

    #[test]
    fn well_formed_multi_line_overlay() {
        let mut table = [empty_entry(); MAX_ENTRIES];
        let content = b"PERMS.TAB 0600 0 0\nSHADOW 0600 0 0\nROUNDTR.DAT 0666 0 0\n";
        let count = parse(content, &mut table).unwrap();
        assert_eq!(count, 3);
        assert_eq!(table[0].name(), b"PERMS.TAB");
        assert_eq!(table[0].mode, 0o600);
        assert_eq!(table[0].uid, 0);
        assert_eq!(table[0].gid, 0);
        assert_eq!(table[1].name(), b"SHADOW");
        assert_eq!(table[2].mode, 0o666);
    }

    #[test]
    fn comments_blank_lines_and_surrounding_whitespace_are_skipped() {
        let mut table = [empty_entry(); MAX_ENTRIES];
        let content = b"# FlashOS FAT32 permission overlay\n\n   \n  SHADOW   0600  0   0  \n# trailing comment\n";
        let count = parse(content, &mut table).unwrap();
        assert_eq!(count, 1);
        assert_eq!(table[0].name(), b"SHADOW");
    }

    #[test]
    fn crlf_line_endings_are_tolerated() {
        let mut table = [empty_entry(); MAX_ENTRIES];
        let count = parse(b"SHADOW 0600 0 0\r\nPERMS.TAB 0600 0 0\r\n", &mut table).unwrap();
        assert_eq!(count, 2);
        assert_eq!(table[0].name(), b"SHADOW");
        assert_eq!(table[1].name(), b"PERMS.TAB");
    }

    #[test]
    fn last_line_needs_no_trailing_newline() {
        let mut table = [empty_entry(); MAX_ENTRIES];
        assert_eq!(parse(b"SHADOW 0600 0 0", &mut table), Some(1));
    }

    #[test]
    fn empty_and_comment_only_files_are_valid() {
        let mut table = [empty_entry(); MAX_ENTRIES];
        assert_eq!(parse(b"", &mut table), Some(0));
        assert_eq!(parse(b"# nothing here\n", &mut table), Some(0));
    }

    #[test]
    fn missing_field_rejects_the_whole_overlay() {
        let mut table = [empty_entry(); MAX_ENTRIES];
        assert_eq!(parse(b"SHADOW 0600 0\n", &mut table), None);
        assert_eq!(parse(b"SHADOW 0600\n", &mut table), None);
        assert_eq!(parse(b"SHADOW\n", &mut table), None);
    }

    #[test]
    fn fifth_field_rejects_the_whole_overlay() {
        let mut table = [empty_entry(); MAX_ENTRIES];
        assert_eq!(parse(b"SHADOW 0600 0 0 extra\n", &mut table), None);
    }

    #[test]
    fn one_malformed_line_rejects_the_whole_overlay() {
        let mut table = [empty_entry(); MAX_ENTRIES];
        assert_eq!(
            parse(b"SHADOW 0600 0 0\nPERMS.TAB 9999 0 0\n", &mut table),
            None
        );
    }

    #[test]
    fn non_octal_and_overwide_modes_reject() {
        let mut table = [empty_entry(); MAX_ENTRIES];
        assert_eq!(parse(b"SHADOW 08 0 0\n", &mut table), None);
        assert_eq!(parse(b"SHADOW abc 0 0\n", &mut table), None);
        assert_eq!(parse(b"SHADOW 1777 0 0\n", &mut table), None);
    }

    #[test]
    fn non_decimal_ids_reject() {
        let mut table = [empty_entry(); MAX_ENTRIES];
        assert_eq!(parse(b"SHADOW 0600 root 0\n", &mut table), None);
        assert_eq!(parse(b"SHADOW 0600 0 0x0\n", &mut table), None);
    }

    #[test]
    fn overlong_name_rejects() {
        let mut table = [empty_entry(); MAX_ENTRIES];
        assert_eq!(parse(b"ABCDEFGHI.TXT 0600 0 0\n", &mut table), None);
    }

    #[test]
    fn capacity_overflow_rejects() {
        let mut table = [empty_entry(); 2];
        assert_eq!(
            parse(b"A 0600 0 0\nB 0600 0 0\nC 0600 0 0\n", &mut table),
            None
        );
    }

    #[test]
    fn lookup_is_case_insensitive_and_can_miss() {
        let mut table = [empty_entry(); MAX_ENTRIES];
        let count = parse(b"SHADOW 0600 0 0\nPERMS.TAB 0600 0 0\n", &mut table).unwrap();
        assert_eq!(lookup(&table[..count], b"shadow").unwrap().mode, 0o600);
        assert_eq!(
            lookup(&table[..count], b"perms.tab").unwrap().name(),
            b"PERMS.TAB"
        );
        assert!(lookup(&table[..count], b"roundtr.dat").is_none());
    }

    #[test]
    fn empty_table_always_misses() {
        assert!(lookup(&[], b"shadow").is_none());
    }

    const fn empty_entry() -> Entry {
        Entry {
            name_buf: [0; 12],
            name_len: 0,
            mode: 0,
            uid: 0,
            gid: 0,
        }
    }
}
